apis/core/v1alpha1/resource.go

```diff
@@ -171,13 +171,43 @@ type ProviderSpec struct {
 	CredentialsSecretRef *SecretKeySelector `json:"credentialsSecretRef,omitempty"`
 }
 
-// A ProviderConfigSpec defines the common way to get to the necessary objects
-// to connect to the provider.
+// A ProviderConfigSpec defines the desired state of a provider config. A
+// provider config may embed this type in its spec in order to support standard
+// fields. Provider configs may choose to avoid embedding this type as
+// appropriate, but are encouraged to follow its conventions.
 type ProviderConfigSpec struct {
-	// CredentialsSecretRef references a specific secret's key that contains
-	// the credentials that are used to connect to the provider.
+	// Credentials required to authenticate to this provider.
+	Credentials ProviderCredentials `json:"credentials"`
+}
+
+// A CredentialsSource is a source from which provider credentials may be
+// acquired.
+type CredentialsSource string
+
+const (
+	// CredentialsSourceNone indicates that a provider does not require
+	// credentials.
+	CredentialsSourceNone CredentialsSource = "None"
+
+	// CredentialsSourceSecret indicates that a provider should acquire
+	// credentials from a secret.
+	CredentialsSourceSecret CredentialsSource = "Secret"
+
+	// CredentialsSourceNative indicates that a provider should acquire
```

Comment:
As listed here, there are many different auth methods and they are all very specific to each provider. Some could even require more details we don't anticipate right now, like a role ID or different types of _Native_ authentication. I know they can always make additions to this in their provider but by giving a source type, this struct is authoritative over other fields and trying to cover them all from a central place doesn't really feel future-proof to me.

In fact, I think it's a losing game to try to standardize on a `ProviderConfig` format. I would even push for having only some interfaces here in crossplane-runtime (to cover `ProviderConfigUsage` machinery) and leave the whole implementation to providers. Since it's only one kind per provider, maintenance effort would be negligible compared to all other inline structs we have, which usually have N implementations per provider.

Reply:
> I think it's a losing game to try to standardize on a ProviderConfig format.

It's not a zero sum game. :) I'm not suggesting that all providers _must_ embed this struct; just offering it if they would like to. In my opinion there's value in doing is. It provides:

* A consistent default for providers with simple credential need - e.g. just secrets.
* A consistent way to add a very common field to all providers, if we ever want to do so.
* A reference example for suggested names and docstrings for providers that want to do something different.

I think the fact that this type would be suitable for every extant Crossplane provider speaks to the fact that it's likely to be useful in a lot of cases. If and when providers appear that have different use cases, they could DIY but be guided by this struct; for example perhaps a provider would use the "standard" `source: Native" but implement its own `spec.credentials.native` options.

Reply:
FWIW, I think we should proceed with adding this struct, but if we didn't I would still advocate for the the API changes it contains to be manually made to every provider. I suspect doing so would result in avoidable, unnessaccary divergence and inconsistency in ProviderConfig APIs over time.

Reply:
> I think the fact that this type would be suitable for every extant Crossplane provider speaks to the fact that it's likely to be useful in a lot of cases.

It's only the secret that's common though. For example, if I implement this functionality in provider-aws, I would not use `Native` as source name. I'd go for `IRSA` or `EC2Metadata`, or `WorkloadIdentity` for GCP. `Native` gives clue for all of them but doesn't really cover any of them.

> I suspect doing so would result in avoidable, unnessaccary divergence and inconsistency in ProviderConfig APIs over time.

It's one `kind` in a provider and a set of authentication methods which are already divergent due to nature of cloud authentication. To have a pattern consistency, we can refer to Kubernetes API Patterns and `ProviderConfig` implementation of AWS that has this source pattern with its own source types so that other providers can implement theirs similarly and that'd be pretty much it. It isn't like managed resource patterns that we do everything we can to standardize because of all the common machinery and number of managed resources in each provider. So, I disagree that the divergence is unnecessary.